src/test/util_tests.cpp

```diff
@@ -122,6 +122,11 @@ BOOST_AUTO_TEST_CASE(util_check)
 
     // Check nested Asserts
     BOOST_CHECK_EQUAL(Assert((Assert(x).test() ? 3 : 0)), 3);
+
+    // Check -Wdangling-gsl does not trigger when copying the int. (It would
+    // trigger on "const int&")
+    const int nine{*Assert(std::optional<int>{9})};
```

Comment:
ACK fa3ea81c3e7d962715788ab5525958a532c51414

Cherry-picked this commit onto the Erlay branch I'm testing, rebased to master, Clang 15 debug build clean, then added the reference `&` to provoke the error.
```
  CXX      test/test_bitcoin-util_tests.o
test/util_tests.cpp:128:29: error: object backing the pointer will be destroyed at the end of the full-expression [-Werror,-Wdangling-gsl]
    const int& nine{*Assert(std::optional<int>{9})};
                            ^~~~~~~~~~~~~~~~~~~~~
./util/check.h:75:50: note: expanded from macro 'Assert'
#define Assert(val) inline_assertion_check<true>(val, __FILE__, __LINE__, __func__, #val)
                                                 ^~~
1 error generated.
```